tools/build_langserver/langserver/analyzer.go

```diff
@@ -1,3 +1,199 @@
 package langserver
 
+import (
+	"core"
+	"parse"
+	"strings"
+
+	"parse/asp"
+	"tools/build_langserver/lsp"
+	"io/ioutil"
+	"fmt"
+)
+
 // TODO(bnmetrics): This file should contain functions to retrieve builtin and custom definitions of build defs
+
+// Analyzer is a wrapper around asp.Analyzer
+// This is being loaded into a handler on initialization
+type Analyzer struct {
+	parser *asp.Parser
+	BuiltIns map[string]*RuleDef
+}
+
+// RuleDef is a wrapper around asp.FuncDef,
+// it also includes a Header(function definition)
+// And Argument map stores the name and the information of the arguments this rule has
+type RuleDef struct {
+	*asp.FuncDef
+	Header string
+	ArgMap map[string]*Argument
+}
+
+// Argument is a wrapper around asp.Argument,
+// this is used to store the argument information for specific rules,
+// and it also tells you if the argument is required
+type Argument struct {
+	*asp.Argument
+	definition string
+	required bool
+}
+
+// Identifier is a wrapper around asp.Identifier
+// Including the starting line and the ending line number
+type Identifier struct {
+	*asp.IdentStatement
+	StartLine int
+	EndLine int
+}
+
+func newAnalyzer() *Analyzer {
+	state := core.NewDefaultBuildState()
+	parser := parse.GetParserWithBuiltins(state)
+
+	a := &Analyzer{
+		parser:parser,
+	}
+	a.BuiltInsRules()
+
+	return a
+}
+
+// BuiltInsRules gets all the builtin functions and rules as a map, and store it in Analyzer.BuiltIns
+func (a *Analyzer) BuiltInsRules() {
```

Comment:
This should probably be private too - you only use it when initialising an Analyzer